### PR TITLE
fix/systemVM-deploy-sc

### PR DESCRIPTION
**Pull request by sasurobert**
fix delegation manager deploy of another smart contract.
added invariants for balance delta to not be negative in case of cross shard accounts

### CODE REVIEW

File: integrationTests/vm/delegation/delegation_test.go [hunk 1]
```diff
@@ -114,6 +114,19 @@ func TestDelegationSystemSCWithValidatorStatistics(t *testing.T) {
 	for i := 0; i < len(balancesAfterClaimRewards); i++ {
 		assert.True(t, balancesAfterClaimRewards[i].Cmp(balancesBeforeClaimRewards[i]) > 0)
 	}
+
+	baseIssuingCost := big.NewInt(100)
```
Comment (iulianpascalau) [inline integrationTests/vm/delegation/delegation_test.go:118]: :+1:

File: process/smartContract/process_test.go [hunk 2]
```diff
@@ -1581,6 +1581,10 @@ func TestScProcessor_processSCOutputAccountsNotInShard(t *testing.T) {
 
 	_, _, err = sc.processSCOutputAccounts(&vmcommon.VMOutput{}, vmcommon.DirectCall, outputAccounts, tx, []byte("hash"))
 	require.Nil(t, err)
+
+	outacc1.BalanceDelta = big.NewInt(-50)
```
Comment (iulianpascalau) [inline process/smartContract/process_test.go:1585]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — DISMISSED** [review]

File: process/scToProtocol/stakingToPeer.go [hunk 3]
```diff
@@ -177,11 +177,14 @@ func (stp *stakingToPeer) UpdateProtocol(body *block.Body, nonce uint64) error {
 		var data []byte
 		data = stp.getStorageFromAccount(stakingSCAccount, blsPubKey)
 		// no data under key -> peer can be deleted from trie
-		if len(data) == 0 {
+		var existingAcc state.AccountHandler
+		existingAcc, err = stp.peerState.GetExistingAccount(blsPubKey)
+		if len(data) == 0 && !check.IfNil(existingAcc) && err != nil {
```
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:182]: maybe you can extract this in one variable

File: process/scToProtocol/stakingToPeer.go [hunk 4]
```diff
@@ -177,11 +177,14 @@ func (stp *stakingToPeer) UpdateProtocol(body *block.Body, nonce uint64) error {
...
 			err = stp.peerState.RemoveAccount(blsPubKey)
 			if err != nil {
 				log.Debug("staking to protocol RemoveAccount", "error", err, "blsPubKey", hex.EncodeToString(blsPubKey))
 			}
+			log.Debug("remove account from validator statistics", "blsPubKey", blsPubKey)
```
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:187]: Add continue after line 185, otherwise this log will be displayed also when RemoveAccount will fail

**Review by SebastianMarian — DISMISSED** [review]

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/validator.go [hunk 5]
```diff
@@ -1144,9 +1144,11 @@ func (v *validatorSC) unBond(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
 
 	validatorConfig := v.getConfig(v.eei.BlockChainHook().CurrentEpoch())
 	totalUnBond := big.NewInt(0).Mul(validatorConfig.NodePrice, big.NewInt(int64(len(unBondedKeys))))
-	totalUnBond, returnCode = v.unBondTokensFromRegistrationData(registrationData, totalUnBond)
-	if returnCode != vmcommon.Ok {
-		return returnCode
+	if len(unBondedKeys) > 0 {
+		totalUnBond, returnCode = v.unBondTokensFromRegistrationData(registrationData, totalUnBond)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1148]: what happened if we have returned here directly?

File: process/transaction/shardProcess.go [hunk 6]
```diff
@@ -394,7 +395,7 @@ func (txProc *txProcessor) checkIfValidTxToMetaChain(
 	if txProc.flagMetaProtection.IsSet() {
 		// additional check
 		if tx.GasLimit < txProc.economicsFee.ComputeGasLimit(tx)+core.MinMetaTxExtraGasCost {
-			return process.ErrInvalidMetaTransaction
+			return fmt.Errorf("%w: not enough gas", process.ErrInvalidMetaTransaction)
```
Comment (iulianpascalau) [inline process/transaction/shardProcess.go:398]: is this still backwards compatible?

File: vm/systemSmartContracts/validator_test.go [hunk 7]
```diff
@@ -2040,41 +2040,119 @@ func TestValidatorStakingSC_ExecuteStakeUnStakeReturnsErrAsNotEnabled(t *testing
...
+	arguments := CreateVmContractCallInput()
+	arguments.CallerAddr = caller
+	arguments.Function = "unBond"
+	arguments.Arguments = [][]byte{blsPubKey}
+
+	retCode := validatorSc.Execute(arguments)
+	assert.Equal(t, vmcommon.Ok, retCode)
+	assert.True(t, strings.Contains(eei.returnMessage, "unBond is not possible"))
+	assert.True(t, strings.Contains(eei.returnMessage, "unBond period did not pass"))
+	assert.True(t, len(eei.storageUpdate["staking"][string(caller)]) != 0) //should have not removed the account data
```
Comment (sasurobert) [inline vm/systemSmartContracts/validator_test.go:2096]: eei.GetStorageForAddress would have been better

File: process/scToProtocol/stakingToPeer.go (hunk 3 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go]: Done

File: process/scToProtocol/stakingToPeer.go (hunk 4 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:189]: Done

File: vm/systemSmartContracts/validator.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1148]: Maybe some changes in save will not happen

File: process/transaction/shardProcess.go (hunk 6 above)
Comment (sasurobert) [reply process/transaction/shardProcess.go:398]: Yes it is. As this is under meta protection flag

Comment (iulianpascalau) [reply process/transaction/shardProcess.go:398]: right, slipped that

**Review by iulianpascalau — DISMISSED** [review]

File: vm/systemSmartContracts/delegation.go [hunk 8]
```diff
@@ -561,6 +561,16 @@ func (d *delegation) modifyTotalDelegationCap(args *vmcommon.ContractCallInput)
 	return vmcommon.Ok
 }
 
+func (d *delegation) checkBLSKeysIfExistsInStakingSC(blsKeys [][]byte) bool {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/delegation.go:564]: checkIfAtLeastOneBLSKeyExistsInStakingSC

Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:564]: would leave as it is

File: vm/systemSmartContracts/delegation.go [hunk 9]
```diff
@@ -1755,6 +1757,40 @@ func (d *delegation) getUserUnBondable(args *vmcommon.ContractCallInput) vmcommo
 	return vmcommon.Ok
 }
 
+func (d *delegation) getUserUndelegatedList(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:1760]: An unit test would have been great

Comment (iulianpascalau) [reply vm/systemSmartContracts/delegation.go:1760]: Added task EN-8525

**Review by iulianpascalau — DISMISSED** [review]

File: vm/systemSmartContracts/delegation.go [hunk 10]
```diff
@@ -1755,6 +1757,40 @@ func (d *delegation) getUserUnBondable(args *vmcommon.ContractCallInput) vmcommo
...
+	for _, fundKey := range delegator.UnStakedFunds {
+		fund, err = d.getFund(fundKey)
+		if err != nil {
+			d.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+
+		d.eei.Finish(fund.Value.Bytes())
+		elapsedNonce := currentNonce - fund.Nonce
+		if elapsedNonce < dConfig.UnBondPeriod {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/delegation.go:1783]: ">=" instead "<"  ?

File: vm/systemSmartContracts/delegation.go [hunk 11]
```diff
@@ -1755,6 +1757,40 @@ func (d *delegation) getUserUnBondable(args *vmcommon.ContractCallInput) vmcommo
...
+			d.eei.Finish(zero.Bytes())
+			continue
+		}
+
+		d.eei.Finish(big.NewInt(0).SetUint64(elapsedNonce).Bytes())
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/delegation.go:1788]: dConfig.UnBondPeriod - elapsedNonce  ?

File: vm/systemSmartContracts/delegation.go (hunk 10 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go]: done

File: vm/systemSmartContracts/delegation.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go]: done

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/delegation.go [hunk 12]
```diff
@@ -1780,12 +1780,13 @@ func (d *delegation) getUserUndelegatedList(args *vmcommon.ContractCallInput) vm
...
 		d.eei.Finish(fund.Value.Bytes())
 		elapsedNonce := currentNonce - fund.Nonce
-		if elapsedNonce < dConfig.UnBondPeriod {
+		if elapsedNonce >= dConfig.UnBondPeriod {
 			d.eei.Finish(zero.Bytes())
 			continue
 		}
 
-		d.eei.Finish(big.NewInt(0).SetUint64(elapsedNonce).Bytes())
+		remainingNonce := dConfig.UnBondPeriod - elapsedNonce
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/delegation.go:1788]: remainingNonces

**Review by SebastianMarian — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**